src/types.ts

```diff
@@ -157,13 +160,29 @@ export interface Service {
 	user?: string;
 	userns_mode?: string;
 
-	volumes?: ListOfUniqueItems<string>;
+	volumes?: ListOfUniqueItems<string | VolumeRef>;
 	volume_driver?: string;
 	volumes_from?: ListOfUniqueItems<string>;
 
 	working_dir?: string;
 }
 
+export interface VolumeRef {
```

Comment:
dockerode wants volume references as an array of strings: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/dockerode/index.d.ts#L1013

so technically were making things more difficult for downstream by normalizing to `VolumeRef` if they just want to pass those values along. maybe we want to expose a helper to turn a `VolumeRef` into it's string representation?

Reply:
There are two considerations:

- what input is considered acceptable?
- what kind of data we can push downstream?

As far as (1) goes, ideally we'd be able to accept anything schema v2.1 (or v2.4 with this PR) supports.

As far as (2) goes, my comment above hopefully explains why we don't really have an option -- unless you see a way to push breaking runtime config changes downstream.

In my opinion, we should be able to accept a VolumeRef but *always* normalise it to a list of strings, as expected downstream, and consider any limitations this may impose -- eg. can we consistently transform *every* VolumeRef to a string, and if not, how should these cases be handled?

Reply:
a `VolumeRef` is basically one structure to describe all kinds of mounts on a container, unifying `volumes`(strings) and `tmpfs` into one. we can just do what you suggest and make this a compatible change